compiler/rustc_const_eval/src/interpret/intrinsics.rs

```diff
@@ -573,6 +601,106 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+        let c = ImmTy::from_scalar(c.to_scalar(), l.layout);
+
+        let (val, overflowed2) = self.binary_op(mir_op, &val, &c)?.to_scalar_pair();
+
+        let overflowed1 = overflowed1.to_bool()?;
+        let overflowed2 = overflowed2.to_bool()?;
+
+        let overflowed = Scalar::from_bool(if l.layout.abi.is_signed() {
+            overflowed1 != overflowed2
+        } else {
+            overflowed1 | overflowed2
+        });
+
+        interp_ok((val, overflowed))
+    }
+
+    pub fn mul_double_add2(
+        &self,
+        l: &ImmTy<'tcx, M::Provenance>,
+        r: &ImmTy<'tcx, M::Provenance>,
+        c1: Option<&ImmTy<'tcx, M::Provenance>>,
+        c2: Option<&ImmTy<'tcx, M::Provenance>>,
+    ) -> InterpResult<'tcx, (Scalar<M::Provenance>, Scalar<M::Provenance>)> {
+        assert_eq!(l.layout.ty, r.layout.ty);
+        assert_matches!(l.layout.ty.kind(), ty::Int(..) | ty::Uint(..));
+
+        let is_signed = l.layout.abi.is_signed();
+        let size = l.layout.size;
+        let bits = size.bits();
+        let l = l.to_scalar_int()?;
+        let r = r.to_scalar_int()?;
+
+        interp_ok(if is_signed {
+            let l = l.to_int(size);
+            let r = r.to_int(size);
+            let c1 = c1.map_or(interp_ok(0), |c1| interp_ok(c1.to_scalar_int()?.to_int(size)))?;
+            let c2 = c2.map_or(interp_ok(0), |c2| interp_ok(c2.to_scalar_int()?.to_int(size)))?;
+            if bits == 128 {
+                #[cfg(bootstrap)]
+                {
+                    let _ = (l, r, c1, c2);
+                    unimplemented!()
+                }
+                #[cfg(not(bootstrap))]
+                {
+                    let (lo, hi) = l.carrying2_mul(r, c1, c2);
+                    let lo = Scalar::from_uint(lo, size);
+                    let hi = Scalar::from_int(hi, size);
+                    (lo, hi)
+                }
+            } else {
+                let prod = l * r + c1 + c2;
+                let lo = Scalar::from_int(prod, size);
+                let hi = Scalar::from_int(prod >> size.bits(), size);
+                (lo, hi)
+            }
+        } else {
+            let l = l.to_uint(size);
+            let r = r.to_uint(size);
+            let c1 = c1.map_or(interp_ok(0), |c1| interp_ok(c1.to_scalar_int()?.to_uint(size)))?;
+            let c2 = c2.map_or(interp_ok(0), |c2| interp_ok(c2.to_scalar_int()?.to_uint(size)))?;
+            if bits == 128 {
+                #[cfg(bootstrap)]
+                {
+                    let _ = (l, r, c1, c2);
+                    unimplemented!()
+                }
+                #[cfg(not(bootstrap))]
+                {
+                    let (lo, hi) = l.carrying2_mul(r, c1, c2);
+                    let lo = Scalar::from_uint(lo, size);
+                    let hi = Scalar::from_uint(hi, size);
+                    (lo, hi)
+                }
+            } else {
+                let prod = l * r + c1 + c2;
```

Comment:
Please add a comment explaining why this does not cause overflow. Also please use `strict_` operations.

Reply:
You are multiplying two 64bit numbers here, which results in a 128bit number. Then you add more stuff. `u64::MAX * u64::MAX` is fairly close to `u128::MAX`, and then we add stuff... why can't this overflow? I have no intuition for this, so it definitely needs comments.

Reply:
It turns out that `u64::MAX * u64::MAX + u64::MAX + u64::MAX` is actually `u128::MAX`, which is the principle behind allowing up to two carries for double-wide multiplication.

But yes, I'll add comments.